disable clicks on anchor Buttons

jquense opened the pull request:
closes #1415

taion commented:
Can we move this down to `<SafeAnchor>`? See also https://github.com/taion/react-bootstrap/tree/anchorProps – we might want to pass through `disabled` to `<SafeAnchor>` in these cases.

jquense commented:
sure, tho I have to admit I don't really understand what SafeAnchor is doing?

taion commented:
I have no idea.

taion commented:
Pointed at the original reported issue.

jquense commented:
OK...so It looks like the point is to address exactly these sort of issues where `href`-less anchors are treated like buttons by default?

It seems like it would make sense to put logic there, except that in the case of a disabled `Button` with a `href` you'd want to prevent navigation and clicks whereas `SafeAnchor` only handles the cases where href is `undefined`? Looking at our uses tho it seems like if a disabled was passed to SafeAnchor, not firing anything would probably be the desired behavior

taion commented:
Yeah, that's the point I'm trying to make. If we make `onClick` on `SafeAnchor` do something like:

```js
onClick = (e) => {
  if (this.props.disabled) {
    e.preventDefault();
    return;
  }

  if (this.props.onClick) {
    this.props.onClick(e);
  }
}
```

jquense commented:
Just updated with a more robust solution. It may be overkill tho. Also Buttons don't use SafeAnchor hence the new component used by both

jquense commented on src/ButtonAnchor.js:
```diff
@@ -0,0 +1,55 @@
+
+    if (disabled) {
+      event.stopPropagation();
+      return;
```
the shotgun approach to preventing/hiding click events. In general `pointer-events: none` handles everything, these are in case its not supported or someone (most likely Safari) decides that its better to ignore that css property for links,

taion commented:
This doesn't actually solve the problem. The issue arises when `href` _is_ set. For e.g. `<Button>`s without an `href`, the element actually is really disabled.

jquense commented:
right hence the seperate ButtonAnchor component used directly by Button...did I break something in SafeAnchor along the way?

taion commented on src/ButtonAnchor.js:
```diff
@@ -0,0 +1,55 @@
+  handleClick(event) {
+    let { disabled, onClick } = this.props;
+
+    event.preventDefault();
```
You can't unconditionally prevent the default here – this will prevent the normal navigation from working.

jquense replied:
yeah that was the intention, if then link is really a link then it shouldn't use this component.

In other words this is trying to encapsulate "Making a `<button>` but with an `<a>` tag for reasons of browsers being weird"

In the case of SafeAnchor use it will never prevent actual navigation because this only gets rendered when there is no `href`

taion replied:
Ahh I see – right, doesn't hit this code. What do you think of merging this logic into `<SafeAnchor>`? Might be nice to remove one level of nesting for checking things out in devtools.

taion replied:
Wait, no, doesn't this get used whenever we have a `<Button href="whatever" />`?

jquense replied:
bah yeah...so there are two cases to address:

1. links that are really buttons
2. real links that just look like buttons

this addresses 1. but kills 2...

I originally had this in SafeAnchor but Button can't use SafeAnchor...or maybe it can. I'll take another look

taion replied:
Just use SafeAnchor in the render anchor code path. The "has no href" branch will be irrelevant there.

jquense replied:
no yeah SafeAnchor is deciding if a link should act like a button which is based on the `href` whereas Button needs it to always treat it like a button regardless of the href.

let me just fix the navigation thing first and see what makes sense then

taion replied:
I'm clearly a bit confused here but I think that's what we want. If a button has a non-trivial `href`, then per http://getbootstrap.com/css/#callout-buttons-anchor-accessibility it should just be a link with button styling and not otherwise get `role="button"` or whatever.

jquense replied:
yeah. The question for whether it should be inside SafeAnchor comes down to this:

For everything else: NavItems, MenuItems, etc should the disabled logic be the same as `<button>` even when there is an `href`? or should that only be the behavior for all Buttons, and links with no hrefs?

taion replied:
We're obnoxiously inconsistent here.

`<MenuItem>` prevents default whenever `props.disabled` is set: https://github.com/react-bootstrap/react-bootstrap/blob/master/src/MenuItem.js#L17-L19 (not sure what preventing the default accomplishes when there's no `href`, though)

`<NavItem>` unconditionally prevents the default so long as `onSelect` is set: https://github.com/react-bootstrap/react-bootstrap/blob/master/src/NavItem.js#L73-L74

jquense replied:
yeah I think I'm for just letting SafeAnchor handle all that...Since there are no disabled semantics for anchors we might as well just follow the input semantics as best we can and call it a day. I can't think of a use case where you'd want navigation to continue will styling something as "disabled"

taion replied:
Makes sense. I have some code earlier for making link handling be consistent on https://github.com/taion/react-bootstrap/commit/67a45851cb6d472f42996c45e16d403e212c15c0 – I think it'd make more sense to build on top of that.

taion replied:
Although we'd also have to amend that implementation to pass along `disabled` to the `<SafeAnchor>` as well.

taion commented on src/ButtonAnchor.js:
```diff
@@ -0,0 +1,55 @@
+  render() {
+    let { href, role, tabIndex, disabled, style } = this.props;
+
+    if (disabled) {
```
Can't we let the TWBS CSS handle this?

jquense replied:
The reason for it is that even tho the css will make it looked disabled it won't prevent focus, for disabled anchors. The `pointer-event: none` will be set on `a.btn.disabled` components by bootstrap css but not for anything else like MenuItems, where we'd also want to keep disabled ones from being navigatable by arrow keys and clicks.

taion commented:
Oh, I see. Do we actively not want this behavior for other components (for both `onClick` and `href`, I mean).

jquense commented:
updated the tests and moved everything to SafeAnchor..I think its not doing the right thing.

We should, as you said also remove that functionality from places where its otherwise specified

taion commented:
I think this is right – what do you think might be wrong?

jquense commented:
just a typoe, sorry i DO think its working correctly

taion commented:
Ah, okay. LGTM then.
